In [2]:
# Imports and pip installations (if needed)
# used 383gpt to figure out installation/imports
import pandas as pd

# Part 1: Load the dataset

In [3]:
# load the given datasets
categorical_data = pd.read_csv('/workspaces/assignment-3-data-cleaning-and-preprocessing-<redacted>/chronic_kidney_disease_categorical.csv')
numerical_data = pd.read_csv('chronic_kidney_disease_numerical.csv')

# print the data
print(categorical_data) #430 rows 14 columns
print(numerical_data) #420 rows 12 cols


     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

The Chronic Kidney Disease (CKD) dataset contains detailed medical data from 400 patients (listed as instances on the website). The goal of this dataset is to understand the factors related to CKD and use them to predict whether a person has CKD or not. By preprocessing this data, we can prepare it for machine learning models that can be used to predict the likelihood of CKD based on the provided medical features. 

The medical details, lab results, and symptoms that are collected from patients are the features that we are using to predict whether someone has CKD (the label). The label is represented as ckd or notckd according to the website. Theses represent someone either having CKD or not having CKD.

The website uses the term "nominal" to describe categorical features. There are both categorical and numerical features. Here are some examples:

Numerical features:
- Age
- Blood Pressure
- White Blood Cell count

Categorical features: 
- Specific Gravity
- Albumin
- Sugar

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [4]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows

# check for dupe rows
num_dupes = numerical_data.duplicated()

# print total number
print(f"There is a total of {num_dupes.sum()} duplicate rows in the numerical dataset.")

# Drop these duplicate rows
dupeless_num_data= numerical_data.drop_duplicates()


# Repeat the same for categorical dataset. Print the duplicate rows and drop them

# check for dupe rows
cat_dupes = categorical_data.duplicated()

# print total number
print(f"There is a total of {cat_dupes.sum()} duplicate rows in the categorical dataset.")

# Drop these duplicate rows
dupeless_cat_data= categorical_data.drop_duplicates()

There is a total of 30 duplicate rows in the numerical dataset.
There is a total of 30 duplicate rows in the categorical dataset.


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

I don't think all the unique ids are present in both datasets because after I merged the data using pd.merge() based on the common identifier "unique_id" and checked for rows where all the columns from dupeless_num_data are NaN, I saw that there were 10 rows meeting this criteria. Similarly, I checked for rows where all the columns from dupeless_cat_data are NaN, and I found that none were missing, so the categorical dataset has all the unique_ids.

Since I used outer join, a unique_id row that is missing from the numerical data but is present in the categorical data will have NaN values for the columns from the numerical dataset. After printing the combined dataset, I know that the first row is present in both datasets, and after printing the missing_ids, I know that the 48th row (index 47) is missing from the numerical dataset. I have printed both to show the difference.

In [5]:
# to see full output, view as scrollable element
combined_data = pd.merge(dupeless_num_data, dupeless_cat_data, on='unique_id', how='outer')

print("First Five Rows of Combined Dataset:")
print(combined_data.head()) # 5 rows

missing_in_num_data = combined_data[dupeless_num_data.columns.difference(['unique_id'])].isna().all(axis=1)
missing_in_cat_data = combined_data[dupeless_cat_data.columns.difference(['unique_id'])].isna().all(axis=1)

missing_ids = combined_data[missing_in_num_data | missing_in_cat_data]

#print(missing_in_cat_data.sum()) 0
#print(missing_in_num_data.sum()) 10
print("\nMissing IDs:")
print(missing_ids)
print("\nRow that is present in both datasets:") 
print(combined_data.iloc[0])
print("\nRow that is missing from numerical dataset, but is present in categorical dataset:")  
print(combined_data.iloc[47])  


First Five Rows of Combined Dataset:
   unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0     102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1     103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2     104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3     105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4     107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   

         pc         pcc          ba  htn   dm cad appet   pe  ane  Target  
0  abnormal  notpresent  notpresent  yes   no  no  poor  yes  yes     ckd  
1    normal  notpresent  notpresent   no   no  no  good   no   no  notckd  
2    normal     present  notpresent   no  yes  no  good   no   no     ckd  
3    normal  notpresent  notpresent  yes  yes  no  good   no   no     ckd  
4    normal  notpresent  notpresent   no   no  no  good   no   no  notckd  

[5 rows x 25 columns]

Missing IDs:
     un

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [6]:
# Calculate the percentage of rows that contain atleast one missing value (used 383gpt)
missing_rows = combined_data.isna().any(axis=1)  
missing_percentage = missing_rows.mean() * 100  
# Print %
print(f"Percentage of rows with at least one missing value: {missing_percentage:.2f}%")
# Drop these rows from the dataset
cleaned_data = combined_data.dropna()
# Print the Dataset
print(f'Data with missing rows dropped: {cleaned_data}')

Percentage of rows with at least one missing value: 61.75%
Data with missing rows dropped:      unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0

## Part 3.4: Sort the dataset according to the Labels

In [7]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting

sorted_data_by_target = cleaned_data.sort_values(by='Target')
# Print the dataset
print(sorted_data_by_target)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
103     337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
105     343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
108     349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
110     353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
171     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
43      203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
169     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
54      228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

Thr correct pandas function to do this is pd.get_dummies().

In [8]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
cat_sorted_data = pd.get_dummies(sorted_data_by_target, columns=categorical_columns, prefix=categorical_columns)
print(cat_sorted_data)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
103     337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
105     343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
108     349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
110     353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
171     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
43      203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
169     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
54      228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

In this situation, we'd have a different column for each condiment and the values would be True/False if they are present or not.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset


In [9]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for col in numerical_columns:
    mean = cat_sorted_data[col].mean()
    deviation = cat_sorted_data[col].std()
    upper = mean + 3 * deviation
    lower = mean - 3 * deviation
    cat_sorted_data = cat_sorted_data[(cat_sorted_data[col] < upper) & (cat_sorted_data[col] > lower)]

print(cat_sorted_data)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
105     343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
131     397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
149     438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
163     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
164     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
171     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0  ...   
43      203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0  ...   
169     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0  ...   
54      228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0  ...   
396     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [10]:
# Normalize the all Numerical Attributes in the dataset.

for col in numerical_columns:
    min_value = cat_sorted_data[col].min()
    max_value = cat_sorted_data[col].max()
    cat_sorted_data[col] = (cat_sorted_data[col] - min_value) / (max_value - min_value)

# Print the normalized DataFrame
print(cat_sorted_data)

     unique_id       age    bp       bgr        bu        sc       sod  \
105     343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
131     397388  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
149     438182  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
163     474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
164     475200  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
..         ...       ...   ...       ...       ...       ...       ...   
171     486397  0.189189  0.25  0.227468  0.222222  0.074074  0.500000   
43      203466  0.216216  0.00  0.115880  0.052288  0.098765  0.600000   
169     482105  0.513514  0.00  0.107296  0.235294  0.012346  0.600000   
54      228849  0.702703  0.50  0.128755  0.261438  0.098765  0.666667   
396     996076  0.689189  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo       pcv  ...  cad_no  cad_yes  appet_good  \
105  0.793103  0.000000  0.032258  ...  

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

Columns may be removed if they are irrelevant for prediction. ID columns can be dropped as they don't correlate with the outcome variable and would not help in generalization. We only use ID columns to separate patient data. 

I feel that all numerical columns are relevant as they provide useful information about what factors cause CKD if we see repeated correlation between a numerical factor and the presence of CKD. I think the categorical columns can help identify what symptoms are present when a patient has CKD so I'd like to keep those as well. Dividing them into dummies also helps separate each situation so I would like to keep the dummy columns. The only dummy column I do not want is the Target_not_ckd because I feel that it is redundant.

In [22]:
#Remove that column
dropping = ['unique_id', 'Target_notckd']
final = cat_sorted_data.drop(columns=dropping)
# Print the dataset
print(final)

          age    bp       bgr        bu        sc       sod       pot  \
105  0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
131  0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.206897   
149  0.905405  0.50  0.785408  0.862745  0.518519  0.600000  1.000000   
163  0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
164  0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.586207   
..        ...   ...       ...       ...       ...       ...       ...   
171  0.189189  0.25  0.227468  0.222222  0.074074  0.500000  0.310345   
43   0.216216  0.00  0.115880  0.052288  0.098765  0.600000  0.482759   
169  0.513514  0.00  0.107296  0.235294  0.012346  0.600000  0.448276   
54   0.702703  0.50  0.128755  0.261438  0.098765  0.666667  0.206897   
396  0.689189  0.00  0.150215  0.254902  0.098765  1.000000  0.620690   

         hemo       pcv      wbcc  ...  dm_yes  cad_no  cad_yes  appet_good  \
105  0.000000  0.032258  0.395161  ...    Tr

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [24]:
# Export the dataframe to a new csv file
final.to_csv('final.csv', index=False) # omitting indices to maintain removal of unnecessary info

# Export the DataFrame to a new JSON file
final.to_json('final.json', orient='records', lines=True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [28]:
# getting 15 rows:

print(final.head(15))

          Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
105  0.743243            0.50              0.442060    0.901961   
131  0.770270            1.00              0.901288    0.163399   
149  0.905405            0.50              0.785408    0.862745   
163  0.202703            0.75              0.158798    0.196078   
164  0.783784            1.00              0.399142    0.287582   
168  0.729730            0.00              0.935622    0.169935   
187  0.675676            0.75              0.253219    0.633987   
395  0.851351            0.25              0.618026    0.562092   
206  0.540541            0.00              0.399142    0.535948   
390  0.756757            0.25              0.223176    0.209150   
381  0.662162            0.50              0.618026    0.411765   
362  0.783784            0.00              0.725322    0.313725   
330  0.878378            0.00              0.206009    0.751634   
323  0.878378            0.25              0.639485    0.47058

### Paste the markdown table here
|     age |    bp |      bgr |       bu |       sc |       sod |      pot |     hemo |      pcv |      wbcc |      rbcc |   al |   su | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd |
|---------|-------|----------|----------|----------|-----------|----------|-----------|-----------|-----------|-----------|------|------|---------------|------------|-------------|-----------|----------------|-------------|----------------|------------|--------|---------|-------|--------|--------|---------|-------------|-------------|-------|--------|--------|---------|-------------|
| 0.743243|  0.50 | 0.442060 | 0.901961 | 0.432099 | 0.500000  | 0.793103 | 0.000000  | 0.032258  | 0.395161  | 0.057143  |  2.0 |  0.0 |        True   |      False |      True   |    False  |           True |      False  |           True |      False  |  False |   True  |  False|   True |  False |   True  |      False   |       True   |  False|   True |   False |   True |    True     |
| 0.770270|  1.00 | 0.901288 | 0.163399 | 0.345679 | 0.766667  | 0.206897 | 0.524752  | 0.548387  | 0.443548  | 0.342857  |  2.0 |  2.0 |       False   |       True |     False   |    True   |           True |      False  |          False |       True  |  False |   True  |  True |  False |  False |   True  |       True    |        False  |   True |   False|  True   |  False |    True     |
| 0.905405|  0.50 | 0.785408 | 0.862745 | 0.518519 | 0.600000  | 1.000000 | 0.277228  | 0.322581  | 0.233871  | 0.371429  |  2.0 |  0.0 |        True   |      False |      True   |    False  |           True |      False  |           True |      False  |  False |   True  |  False|   True |  False |   True  |      True    |        False  |   True |   False|  True   |  False |    True     |
| 0.202703|  0.75 | 0.158798 | 0.196078 | 0.160494 | 0.166667  | 0.206897 | 0.059406  | 0.000000  | 0.653226  | 0.257143  |  4.0 |  0.0 |       False   |       True |      True   |    False  |          False |       True  |          False |       True  |   True |  False  |  True | False  |  True  |  False  |      True     |        False  |  True |   False|  False |  True  |    True     |
| 0.783784|  1.00 | 0.399142 | 0.287582 | 0.839506 | 0.666667  | 0.586207 | 0.019802  | 0.096774  | 0.258065  | 0.114286  |  4.0 |  2.0 |        True   |      False |      True   |    False  |           True |      False  |           False|       True  |  False |   True  | False |  True  |  True  |  False  |      True     |        False  |  False|   True |  True  |  False |    True     |
| 0.729730|  0.00 | 0.935622 | 0.169935 | 0.160494 | 0.333333  | 0.034483 | 0.019802  | 0.064516  | 0.879032  | 0.000000  |  3.0 |  1.0 |       False   |       True |      True   |    False  |          False |       True  |           True |      False  |  False |   True  |  True |  False |  True  |  False  |      False    |        True   |  True |   False|  False |  True  |    True     |
| 0.675676|  0.75 | 0.253219 | 0.633987 | 0.777778 | 0.366667  | 0.655172 | 0.138614  | 0.193548  | 0.169355  | 0.114286  |  2.0 |  0.0 |        True   |      False |      True   |    False  |           True |      False  |           True |      False  |  False |   True  |  True |  False |  True  |  False  |      True     |        False  |   True|   False|  True  |  False |    True     |
| 0.851351|  0.25 | 0.618026 | 0.562092 | 0.728395 | 0.000000  | 0.344828 | 0.168317  | 0.161290  | 0.580645  | 0.085714  |  4.0 |  3.0 |       False   |       True |      True   |    False  |          False |       True  |          False |       True  |  False |   True  | False |  True  |  False |   True  |      True     |        False  |  False|   True |  False |   True |    True     |
| 0.540541|  0.00 | 0.399142 | 0.535948 | 0.358025 | 0.700000  | 0.379310 | 0.207921  | 0.161290  | 0.830645  | 0.057143  |  1.0 |  0.0 |       False   |       True |     False   |    True   |           True |      False  |           True |      False  |  False |   True  | False |  True  |  False |   True  |      True     |        False  |   True|   False|  True  |  False |    True     |
| 0.756757|  0.25 | 0.223176 | 0.209150 | 0.160494 | 0.533333  | 0.620690 | 0.485149  | 0.516129  | 0.290323  | 0.257143  |  3.0 |  0.0 |       False   |       True |      True   |    False  |           True |      False  |           True |      False  |  False |   True  | False |  True  |  True  |  False  |      True     |        False  |   True|   False|  True  |  False |    True     |
| 0.662162|  0.50 | 0.618026 | 0.411765 | 0.432099 | 0.566667  | 0.689655 | 0.316832  | 0.354839  | 0.250000  | 0.200000  |  3.0 |  1.0 |       False   |       True |      True   |    False  |          False |       True  |          False |       True  |  False |   True  | False |  True  |  True  |  False  |      True     |        False  |  False|   True |  True  |  False |    True     |
| 0.783784|  0.00 | 0.725322 | 0.313725 | 0.481481 | 0.566667  | 0.862069 | 0.178218  | 0.193548  | 0.258065  | 0.114286  |  4.0 |  1.0 |        True   |      False |      True   |    False  |           True |      False  |           False|       True  |  False |   True  | False |  True  |  False |   True  |      True     |        False  |   True|   True |  False |   True |    True     |
| 0.878378|  0.00 | 0.206009 | 0.751634 | 0.604938 | 0.533333  | 0.689655 | 0.366337  | 0.387097  | 0.879032  | 0.371429  |  4.0 |  0.0 |       False   |       True |     False   |    True   |           True |      False  |           True |      False  |  False |   True  | False |  True  |  False |   True  |      True     |        False  |   True|   False|  True  |  False |    True     |
| 0.878378|  0.25 | 0.639485 | 0.470588 | 0.395062 | 0.433333  | 0.517241 | 0.267327  | 0.322581  | 0.104839  | 0.171429  |  3.0 |  0.0 |       False   |       True |      True   |    False  |          False |       True  |          False |       True  |  False |   True  | False |  True  |  True  |  False  |      True     |        False  |  False|   True |  True  |  False |    True     |
| 0.716216|  0.50 | 1.000000 | 0.163399 | 0.111111 | 0.066667  | 0.206897 | 0.267327  | 0.387097  | 0.532258  | 0.371429  |  1.0 |  0.0 |        True   |      False |     False   |    True   |           True |      False  |           True |      False  |  True  |  False  | False |  True  |  True  |  False  |      False    |        True   |  True |   False|  True  |  False |    True     |

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

When I first tried doing this task, 383gpt kept cutting the markdown table off at certain lengths. I then googled why this might happen and I learned that I may either be running into character limitations. So I broke the data down into chunks of 5 rows at a time and tried sending them to 383gpt one at a time. This worked, but it started to change some cells. 

From personal experience, if an LLM is not performing the way I want it to, I find that adding some version of "this is extremely important" to the prompt makes it work better. I think this has to do with creating some stakes. Something I found off of an article on Medium is that asking an LLM if it has any questions about the task I assigned will ensure that there is no miscommunication and I am more likely to get my desired result. 

Combining all of these little improvements, I sent this prompt to 383gpt: 

"I am going to give you data in chunks of 3 rows at a time
create a markdown table for every piece of data i give you
this is extremely important so do not make any errors
Before you start, please ask me any questions you have about this so I can give you more context. Be extremely comprehensive.
if there are no questions, simply say okay, and i will begin giving you data"

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [41]:
# Create a dictionary for renaming columns
rename_dict = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'bgr': 'Blood Glucose Random',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count',
    'rbcc': 'Red Blood Cell Count',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc_abnormal': 'Red Blood Cells (Abnormal)',
    'rbc_normal': 'Red Blood Cells (Normal)',
    'pc_abnormal': 'Pus Cell (Abnormal)',
    'pc_normal': 'Pus Cell (Normal)',
    'pcc_notpresent': 'Pus Cell Clumps (Not Present)',
    'pcc_present': 'Pus Cell Clumps (Present)',
    'ba_notpresent': 'Bacteria (Not Present)',
    'ba_present': 'Bacteria (Present)',
    'htn_no': 'Hypertension (No)',
    'htn_yes': 'Hypertension (Yes)',
    'dm_no': 'Diabetes Mellitus (No)',
    'dm_yes': 'Diabetes Mellitus (Yes)',
    'cad_no': 'Coronary Artery Disease (No)',
    'cad_yes': 'Coronary Artery Disease (Yes)',
    'appet_good': 'Appetite (Good)',
    'appet_poor': 'Appetite (Poor)',
    'pe_no': 'Pedal Edema (No)',
    'pe_yes': 'Pedal Edema (Yes)',
    'ane_no': 'Anemia (No)',
    'ane_yes': 'Anemia (Yes)',
    'CKD or Not CKD': 'CKD present'
}

# Apply the renaming to the DataFrame
final.rename(columns=rename_dict, inplace=True)

# Display the updated DataFrame
print(final.head())  # Print the first few rows to verify column names

          Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
105  0.743243            0.50              0.442060    0.901961   
131  0.770270            1.00              0.901288    0.163399   
149  0.905405            0.50              0.785408    0.862745   
163  0.202703            0.75              0.158798    0.196078   
164  0.783784            1.00              0.399142    0.287582   

     Serum Creatinine    Sodium  Potassium  Hemoglobin  Packed Cell Volume  \
105          0.432099  0.500000   0.793103    0.000000            0.032258   
131          0.345679  0.766667   0.206897    0.524752            0.548387   
149          0.518519  0.600000   1.000000    0.277228            0.322581   
163          0.160494  0.166667   0.206897    0.059406            0.000000   
164          0.839506  0.666667   0.586207    0.019802            0.096774   

     White Blood Cell Count  ...  Diabetes Mellitus (Yes)  \
105                0.395161  ...                     True   
131   

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [42]:

result = final.groupby('CKD present').size().reset_index(name='count')


print(result)

   CKD present  count
0        False    111
1         True     22


The `groupby('CKD present')` method groups our data by the `CKD present` column.

The `size()` function counts the number of occurrences in each group.

`reset_index(name='count')` converts the result of using size() back to a DataFrame and names the count column as `count`.